contents/c++/rvalue.md

```diff
@@ -0,0 +1,145 @@
+```
+lvalue = rvalue
+```
+where lvalue (left value) is on the left side of the equation, and rvalue (right value) is on the right.
+lvalue basically refers to values that are addressable, whileas rvalue are temporary objects or values which are used only on the right side of an assignment expression. More details and classification of these 2 values can be found [here](http://www.bogotobogo.com/cplusplus/C11/4_C11_Rvalue_Lvalue.php).
+
+### Using rvalue references
+Rvalue references allows us to distinguish an lvalue from an rvalue. In C++11, we can declare an rvalue reference using the `&&` operator:
+```cpp
+int &&rvalue = 55;
+```
+
+We can also convert an lvalue to a rvalue using the `std::move` function:
+```cpp
+int lvalue = 99;
+int &&rvalue2 = std::move(lvalue); // std::move converts an lvalue to a rvalue
+```
+
+We can also do [function overloading](https://en.wikibooks.org/wiki/Computer_Programming/Function_overloading)
+in order to determine whether the parameters given are lvalues or rvalues:
+
+```cpp
+void print(int& lvalue) { // takes an lvalue
+    std::cout << "lvalue method used";
+}
+
+void print(int&& rvalue) { // take a rvalue
+    std::cout << "rvalue method used";
+}
+
+int main() {
+    int x = 5;
+    print(x); // lvalue method used
+    print(10); // rvalue method used.
+}
+```
+
+Using of these functions overload with rvalue parameters, which take on temporary objects, helps us to write more efficient program using Move Semantics!
+
+### Move Semantics
+The main usage of rvalue references is that it allows us to create *move* constructor and *move* assignments, instead of *copy* constructor and *copy* assignments by default. Since rvalues are typically temporary objects, we can just *move* the value instead of *copying* it, thus helping us to save on memory consumption and improving on the performance on our program as well!
+
+One example implementation of a move constructor and move assignment is:
+```cpp
+Foo(Foo&& other) { // move constructor
+    x = other.x;
+    y = other.y;
+    z = other.z;
+}
+
+Foo& operator=(Foo&& other) { // move assignment
+    x = std::move(other.x);
+    y = std::move(other.y);
+    z = std::move(other.z);
+    return *this;
+}
+```
+
+In this move constructor and assignment, the contents of the `other` parameter is *moved* into the object, and the contents in `other` is destroyed afterwards. No additional memory allocation occurs, and the *move* operation is done quickly by a few assignment of address operations, leading to a more performance and memory efficient program.
+
+Along with Move Semantics in C++11, the [STL library](https://www.geeksforgeeks.org/the-c-standard-template-library-stl/) functions also defines the overloaded *move* functions for its container classes, thus we can take advantage of these Move Semantics by simply supplying rvalues, without the need to redefine the classes ourselves.
```

Comment:
Along with Move Semantics in C++11, the STL library ~~functions also defines the~~ **provides** overloaded move functions for its container classes, thus we can take advantage of these Move Semantics by simply supplying rvalues, without the need to redefine the classes ourselves

Maybe it's better to explain what the container classes here refer to.

Reply:
Yup, that would be better :+1: